File: src/data/courses/dsa/graphs/traversals-basic-properties/detect-cycle-directed/code/java/DetectCycleDirected.java
// ============================================================
// Detect Cycle Directed
// ============================================================
import java.util.*;
class Solution {
    public boolean hasCycle(int n, List<List<Integer>> adj) {
        int[] color = new int[n];
        for (int i = 0; i < n; i++)
            if (color[i] == 0 && dfs(i, adj, color)) return true;
        return false;
    }
    boolean dfs(int node, List<List<Integer>> adj, int[] color) {
        color[node] = 1;
        for (int nb : adj.get(node)) {
            if (color[nb] == 1) return true;
            if (color[nb] == 0 && dfs(nb, adj, color)) return true;
        }
        color[node] = 2;
        return false;
    }
}

File: src/data/courses/dsa/graphs/traversals-basic-properties/detect-cycle-directed/code/python/detect_cycle_directed.py
# ============================================================
# Detect Cycle Directed
# ============================================================
class Solution:
    def hasCycle(self, n, adj):
        color = [0] * n  # 0=white, 1=gray, 2=black
        def dfs(node):
            color[node] = 1  # Gray
            for nb in adj[node]:
                if color[nb] == 1: return True  # Back edge = cycle
                if color[nb] == 0 and dfs(nb): return True
            color[node] = 2  # Black
            return False
        return any(color[i] == 0 and dfs(i) for i in range(n))
